### PR TITLE
Code level metrics

### PR DESCRIPTION
**Pull request by meiao**
### Overview
Adds Code Level Metrics attributes to spans created by @Trace annotations and XML, indicating the class/method that span relates to.
These attributes are code.namespace and code.function.

### CODE REVIEW

File: newrelic-agent/src/test/resources/com/newrelic/agent/config/span_events_clm.yml [hunk 1]
```diff
@@ -0,0 +1,155 @@
+#
+# This file configures the NewRelic Agent.  Place a copy of this file
+# in RAILS_ROOT/config and edit the settings as appropriate.
```
Comment (meiao) [inline newrelic-agent/src/test/resources/com/newrelic/agent/config/span_events_clm.yml:3]: didn't notice this.

File: newrelic-agent/src/test/resources/com/newrelic/agent/config/span_events_clm.yml [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  app_name: Unit Test
+  enable_auto_app_naming: false
+  app_renaming:
+    App1: App12
+    App2: App12
+    App3: App4
+  trim_stats: false
+
+  code_level_metrics:
+    enabled: true
```
Comment (meiao) [inline newrelic-agent/src/test/resources/com/newrelic/agent/config/span_events_clm.yml:20]: This file was copied from span_events.yml and only these 2 lines were added.

File: newrelic-agent/src/main/java/com/newrelic/agent/config/AgentConfig.java [hunk 3]
```diff
@@ -179,6 +179,11 @@ public interface AgentConfig extends com.newrelic.api.agent.Config, DataSenderCo
      */
     ApplicationLoggingConfig getApplicationLoggingConfig();
 
+    /**
+     * Get the Code Level Metrics config.
+     */
+    CodeLevelMetricsConfig getCodeLevelMetricsConfig();
```
Comment (tbradellis) [inline newrelic-agent/src/main/java/com/newrelic/agent/config/AgentConfig.java:185]: I noticed sometimes we have added new config to the interface and some have only been added to the AgentConfigImpl.  I'm just curious for my own knowledge what is the determining factor on whether to include it here.  Also there is a confusing comment on the class which I've never fully understood why it was there:
```
* Adding new methods to this is no longer preferred. Please access new config using the {@link com.newrelic.api.agent.Config#getValue(String)}
 * method with the keyPath being '.' delimited hierarchy of keys, similar to the command line setting override.<br/>. 
```
Does it apply here or is this comment itself outdated?

File: newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    public void enabled() {
+        localProps.put("code_level_metrics.enabled", true);
+        CodeLevelMetricsConfigImpl config = new CodeLevelMetricsConfigImpl(localProps);
+        assertFalse(config.isEnabled());
+    }
+
+    @Test
+    public void canConfigureViaSystemProperty() {
+        Properties properties = new Properties();
+        //default application_logging is true
```
Comment (tbradellis) [inline newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java:44]: I think missed a comment change here.

File: newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+        CodeLevelMetricsConfigImpl config = new CodeLevelMetricsConfigImpl(Collections.emptyMap());
+        assertTrue(config.isEnabled());
+
+    }
+
+    @Test
+    public void canConfigureViaEnvironmentVariables() {
+
+        //default forwarding is false
```
Comment (tbradellis) [inline newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java:61]: same. I think it's a comment from logging.

**Review by tbradellis — APPROVED** [review]
Just saw some comments that could be cleaned up. Had one question but it was for my own understanding.

File: newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java (hunk 4 above)
Comment (meiao) [reply newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java:44]: Right. Thanks.

File: newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java (hunk 5 above)
Comment (meiao) [reply newrelic-agent/src/test/java/com/newrelic/agent/config/CodeLevelMetricsConfigImplTest.java:61]: Right. Thanks.

File: newrelic-agent/src/main/java/com/newrelic/agent/config/AgentConfig.java (hunk 3 above)
Comment (meiao) [reply newrelic-agent/src/main/java/com/newrelic/agent/config/AgentConfig.java:185]: It should be added to the interface so we don't rely on the implementation details. This is specifically useful for testing, where we could have a AgentConfigStub that implements AgentConfig. Then if all the code only uses AgentConfig, then we can easily switch to the stub for testing.

I keep forgetting about that comment.
I followed the Logging configuration, which was the last big piece of configuration we added.
I don't really like that comment, to tell you the truth. I like the fact that you call something like `agentConfig.getLoggingConfig().isEnabled`. Or maybe it is just that I really don't like passing Strings to methods (when enums would be a better choice).

**Pull request merged**